File: mobile/src/utils/upi.ts
export function buildUpiPayUrl(args: {
  vpa: string;
  payeeName: string;
  amount: number | string;
  note?: string;
  currency?: string;
}): string {
  const pa = String(args.vpa || '').trim();
  if (!pa) return '';
  const amount = Number(args.amount || 0);
  if (!Number.isFinite(amount) || amount <= 0) return '';
  const params = new URLSearchParams({
    pa,
    pn: String(args.payeeName || 'Shop').trim() || 'Shop',
    am: amount.toFixed(2),
    cu: args.currency || 'INR',
  });
  if (args.note) params.set('tn', String(args.note).slice(0, 80));
  return `upi://pay?${params.toString()}`;
}

File: backend/apps/common/upi.py
from __future__ import annotations

from decimal import Decimal
from urllib.parse import quote, urlencode


def build_upi_pay_url(
    *,
    vpa: str,
    payee_name: str,
    amount: Decimal | str | float | int,
    note: str = "",
    currency: str = "INR",
) -> str:
    pa = str(vpa or "").strip()
    if not pa:
        return ""
    am = Decimal(str(amount or "0")).quantize(Decimal("0.01"))
    if am <= 0:
        return ""
    params = {
        "pa": pa,
        "pn": str(payee_name or "").strip() or "Merchant",
        "am": f"{am:.2f}",
        "cu": currency or "INR",
    }
    if note:
        params["tn"] = str(note).strip()[:80]
    return f"upi://pay?{urlencode(params, quote_via=quote)}"
